dat/proofread.1
.TH proofread 1 YYYY-MM-DD v0.0.0
.SH NAME
proofread \- removes spaces at EOL and adds a newline at EOF
.SH SYNOPSIS
.nf
\fBproofread\fR [\fB\-l\fR] [\fB\-f\fR] [\fB\-\-dry\-run\fR[\fB=\fR(\fBfull\fR|\fBline\fR)]] [\fB\-\fR]
\fBproofread\fR [\fB\-l\fR] [\fB\-f\fR] [\fB\-\-dry\-run\fR[\fB=\fR(\fBfull\fR|\fBline\fR)]] [\fB\-\-mute\fR] \fB+\fR [\fB\-\-\fR] \fI<file>\fR\&...
\fBproofread\fR [\fB\-l\fR] [\fB\-f\fR] [\fB\-\-keep\fR[\fB=\fI<path>\fR]] [\fB\-\-mute\fR] \fB+\fR [\fB\-\-\fR] \fI<file>\fR\&...
\fBproofread\fR (\fB\-h\fR | \fB\-\-help\fR)
\fBproofread\fR (\fB\-v\fR | \fB\-vv\fR | \fB\-vvv\fR)
.fi
.PP
The options starting with '\-' can be stated in a row, i\&.e\&. \fB\-lf\fR. However, other options must be provided as stated above. Thus, it is wrong to write "\fB\-\-dry\-run line\fR" if it meant "\fB\-\-dry\-run=line\fR".
.PP
It is impossible to mix one form with another. For instance, with the combination of the \fB\-l\fR and \fB\-h\fR options, the program would print an error message.
.SH DESCRIPTION
\fBproofread\fR corrects two kinds of formatting problems: spaces at the end of a line (EOL) and a missing newline at the end of a file (EOF). These EOL and EOF problems can be addressed with the \fB\-l\fR and \fB\-f\fR options, respectively.
.PP
Precisely speaking, a "space" can be either a space(' ') or a tab('\\t'). Also, by a "newline", it means a line feed('\\n'). If a carriage return('\\r') is seen, the program behaves as if the character does not exist.
.PP
In that \fB\-l\fR and \fB\-f\fR are both optional, this program does not inspect its input against the condition not requested. What this implies is that if neither were specified, then it would print its output untouched.
.PP
If the program is fed an empty file, it does nothing.
.SH USAGE
The behavior of this program falls into two categories: a dry\-run and an actual run.
.PP
When this program gets an execution with the \fB\-\-dry\-run\fR option specified, the program prints what it would have fixed in an actual run to the \fBstandard output\fR. Otherwise, it performs an actual run and fix the problems requested for it to handle.
.PP
In the presence of the \fB\-\fR option, it takes its input from the \fBstandard input\fR and prints the result to the \fBstandard output\fR. On the other hand, if the \fB+\fR option is specified, then the program gets the input from the file(s) which are specified after this option and it writes each result to each file. If the \fB\-\-keep\fR option is specified with \fB+\fR, then the input files are kept under a given directory, which defaults to "\fI<install\-path>/bak/\fR" unless provided. Otherwise, the original files are overwritten with the corresponding output. The filenames are separated with a space. Incidentally, it is encouraged to place the \fB\-\-\fR marker between "\fB+\fR" and filenames so as to distinguish options from filenames. Meanwhile, the program defaults to "\fB\-\fR" if neither option is specified.
.PP
Please note that if \fB\-\-keep\fR were specified and the filename had a directory portion, that part would be removed. For example, the following invocation
.RS
.sp
$ proofread --keep=mybackup + -- dir/a.txt
.sp
.RE
would result in "\fImybackup/a.txt\fR" if everything went well.
.PP
If the directory name provided to \fB\-\-keep\fR does not exist, the programs exits with failure. In such a case, the input files remain intact since the program does not directly edit the input file. Instead, the program writes to a temporary file, moves the original file under the specified directory, and gives the temp file the name of the input file. The name of temporary files is made by appending the string "\&.prfrd\&.tmp" to the input file name (including the extension, if any).
.PP
Incidentally, if you are going to make an actual run, especially with the \fB\+\fR option, i.e. "overwriting option", it is highly recommended to make a dry\-run before doing so in order to see what are to be affected.
.PP
If multiple files are provided, each file gets an inspection with the options given in the order filenames are specified.
.PP
An example dry\-run session is shown below:
.RS
.sp
.nf
$ proofread \-lf \-\-dry\-run \- < helloworld\&.c
> proofread \-l \-f \-\-dry\-run=full \-
=================================================
   1|#include <stdio\&.h>
   2|
   3|int main(void) {
   4|   puts("Hello, World!");
   5|   return 0;  3
   6|}
   7|\\n
.fi
.sp
.RE
Here, the program reads the input from stdin due to the \fB\-\fR option. Also, by the \fB\-l\fR option, it tests whether there are no redundant spaces at EOL and by \fB\-f\fR, it checks if the last line ends with a newline. It is observed that line 5 and 6 are problematic in the light of the former having three spaces at EOL and the latter not having a newline at EOF. In order to show the result of this inspection, the program highlights the three spaces with a red background and writes the number of spaces at the end of the line, and it writes "\\n" at the end of the file and highlights it with a green background.
.SH OPTIONS
.IP \fB\-l\fR
Removes unnecessary spaces (if any) before the end of a line.
.IP \fB\-f\fR
Puts a newline before the end of a file if there isn't.
.IP "\fB\-\-dry\-run\fR, \fB\-\-dry\-run=\fR(\fBfull\fR|\fBline\fR)"
Makes a dry\-run. Defaults to "\fBfull\fR" if specified without a value. \fBfull\fR causes the whole file to be printed, whereas \fBline\fR prints only lines with a problem.
.IP \fB\-\fR
Uses stdin as a input.
.IP \fB+\fR
Uses the content of file(s) as a input.
.IP \fB\-\-\fR
Denotes the end of options.
.IP <\fIfile\fR>\&...
Represents filename(s). Note that this does not include directory names.
.IP "\fB\-\-keep\fR, \fB\-\-keep=\fR<\fIpath\fR>"
Keeps the original file under the specified directory. Defaults to "bak/" if the value is not provided.
.IP \fB\-\-mute\fR
Suppresses the rename/remove notification when such actions happen.
.IP "\fB\-h\fR, \fB\-\-help\fR"
Prints the manual you are reading know.
.IP \fB\-v\fR
Given once, prints the version of this program. Given twice, prints the version, the installation path, and the information related to this program. Meanwhile, a question mark hangs over this option being given three times. Let's leave it as the delight of when executed for real. Suffice it to say that you'll like it.
.SH EXAMPLES
.B "* Checking for EOL and EOF Problems"
.RS
.sp
$ proofread \-lf \-\-dry-run=line \- < input\&.txt
.sp
.RE
With \fIinput\&.txt\fR redirected, the program checks whether there are EOL and EOF problems and prints the result of the inspection to the standard output.
.PP
.B "* Checking for Multiple Files"
.RS
.sp
$ proofread \-lf \-\-dry\-run=line + \-\- a\&.txt b\&.txt c\&.txt
.sp
.RE
The program reads the three files and checks for each file whether it has EOL and EOF problems. The result is printed to stdout file by file.
.PP
.B "* Fixing EOL and EOF Problems"
.RS
.sp
$ proofread \-lf + \-\- input\&.txt
.sp
.RE
The program reads one file, which is \fIinput\&.txt\fR, and fixes EOL and EOF problems in that file. Notice that this operation modifies the file.
.PP
.B "* Keeping the Original File under a Specific Directory"
.RS
.sp
$ proofread \-lf \-\-keep=backup/ + \-\- input\&.txt
.sp
.RE
The program reads \fIinput\&.txt\fR and fixes EOL and EOF problems in that file. After the correction, it saves the original file under the "backup/" directory.
.PP
.B "* Setting Up a Pre-commit Hook"
.sp
There is a pre-commit hook which is executed automatically when one executes \fBgit commit\fR. This hook executes \fBproofread\fR on the files which are staged (by a prior \fBgit add\fR) and whose state is modified or newly added. Its behavior differs depending on whether the file has local changes or not. If it has, the hook updates the index only. Otherwise, the hook updates both the index and the worktree.
.PP
In order to use it, one must compile it first and place the executable under the \fI.git/hooks\fR directory of the desired git repository:
.RS
.sp
.nf
$ make hook EXT_LIST='".c", ".h"' SHUTUP=1
$ cp hook/pre-commit \fI<git-repo>/.git/hooks\fR
.fi
.sp
.RE
When making the pre-commit hook, one can control two variables:
.RS
.sp
.IP \fBEXT_LIST\fR
Defines the extensions to apply \fBproofread\fR. The value must be enclosed with a single-quote mark and each extension must be enclosed with a double-quote mark. The extensions are comma\-separated. If not specified, all of the files which meet the above conditions are subject to the application.
.IP \fBSHUTUP\fR
Sets the verbosity of the hook. The value of \fB1\fR means \fBtrue\fR and \fB0\fR, \fBfalse\fR.
.sp
.RE
.PP
If the hook is executed with \fBl\fR option, it prints out the extension list at stdout, exiting with the status code 2.
.SH EXIT STATUS
The program exits with EXIT_SUCCESS on normal status. Otherwise, EXIT_FAILURE.
.SH FILES
.I <file>\&.prfrd\&.tmp
.RS
As mentioned above, it is a temporary file created during the fix.
.RE
.nf
.I \&.prfrd\&.short-status
.I \&.prfrd\&.checkout-index-temp
.I \&.prfrd\&.hash-object-temp
.I \&.merge_file_*
.fi
.RS
Temporary files used by the pre-commit hook. Deleted automatically.
.RE
.I dat/proofread\&.1
.RS
The manual page file. Used when the \fB\-\-help\fR option is specified.
.RE
.I dat/proofread\&.1\&.txt
.RS
The text version of the manual page. The program falls back to this file when \fBman\fR(1) is not available.
.RE
.I dat/version\&.txt
.RS
A text file containing the version of this program.
.RE
.SH NOTES
I took it into account\-\-at least I think so\-\-executing this program under the Windows environment. I tried to compile this program with MSVC and it was observed that the program had no problem, although I didn't make a rigorous examination. I also compiled the pre\-commit hook in the same way and tested it in 'Git for Windows'. Incidentally, it looks like the name of the hook must be \fIpre\-commit.exe\fR, not \fIpre\-commit\fR. Except this affair, the hook program seemed to be fine as well.
.SH BUGS
Under the WSL2 Ubuntu environment, as far as I know, there are no known bugs as of now.
.PP
Under the Windows environment, there are no known bugs as of now.
.PP
When you should find a bug, it would be greatly appreciated if you told me about it.
.SH AUTHOR
Developed by Doohyeon Won (logicseeker@naver\&.com)
.PP
For more information, please visit the GitHub repository: \fIhttps://github\&.com/logic\-finder/proofread\fR.


proofread.c
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include "fatal.h"
#include "strutil.h"
#include "wrapper.h"
#include "argparse.h"
#include "readline.h"
#include "optprocessor.h"
#include "colorcode.h"

/**
 * @brief `proofread` is the main logic of this program, hence the name.
 * @param of the result of argument parsing
 * @param filenames an array of filenames
 * @param keeppath the value of `--keep` option
 */
static void proofread(optflg_t *of, dynarr_t *filenames, const char *keeppath);

int main(int argc, const char *argv[]) {
   optflg_t *of;
   dynarr_t *filenames;
   const char *keeppath;  /* --keep=path */

   of = scalloc(1, sizeof *of);
   filenames = parse_args(of, argv, &keeppath);
   process_opts(of, filenames);
   proofread(of, filenames, keeppath);

   free(of);
   dynarr_destroy(filenames);
   return 0;
}

// `run_dry` gets an invocation when `--dry-run` is specified.
static void run_dry(optflg_t *of, dynarr_t *filenames);
// `run_actual` gets an invocation when it performs real tasks.
static void run_actual(optflg_t *of, dynarr_t *filenames, const char *keeppath);

static void proofread(optflg_t *of, dynarr_t *filenames, const char *keeppath) {
   if (of->drn)
      run_dry(of, filenames);
   else
      run_actual(of, filenames, keeppath);
}

// `apply_dryrun` performs a dry-run action on a file based on `of`.
static void apply_dryrun(FILE *fp, optflg_t *of);

static void run_dry(optflg_t *of, dynarr_t *filenames) {
   char *eolstr = of->eol ? "-l " : "";
   char *eofstr = of->eof ? "-f " : "";
   char *substr = of->lne ? "line " : "full ";
   char *modstr = of->owf ? "+ -- " : "- ";

   // handles + option
   if (of->owf) {
      int idx, len;

      len = dynarr_len(filenames);
      for (idx = 0; idx < len; idx++) {
         char *filename;
         FILE *fp;

         // gets a stream
         filename = *((char **) dynarr_get(filenames, idx));
         fp = sfopen(filename, "r");

         // echoes the enabled options
         if (idx > 0) sputs("");
         fmtwrt("%s> proofread %s%s--dry-run=%s%s%s%s\n",
            Cyellow, eolstr, eofstr, substr, modstr, filename, Creset);
         sputs("=================================================");

         // main logic
         apply_dryrun(fp, of);

         // cleanup
         sfclose(fp);
      }
   }
   // handles the default option
   else {
      // echoes the enabled options
      fmtwrt("%s> proofread %s%s--dry-run=%s%s%s\n",
         Cyellow, eolstr, eofstr, substr, modstr, Creset);
      sputs("=================================================");

      // main logic
      apply_dryrun(stdin, of);
   }
}

// `handle_eoldrn` checkes the eol problem in a dry-run.
static int handle_eoldrn(char *line, int numline, bool dryrun_line);
// `handle_eofdrn` checkes the eof problem in a dry-run.
static bool handle_eofdrn(int numline, int lastmod, bool dryrun_line, bool newline_at_eol);

static void apply_dryrun(FILE *fp, optflg_t *of) {
   int numline, lret, fret, lastmod;
   bool newline_at_eol;
   char *line, *crnt, *prev;

   numline = 1;   /* current line number */
   lret = 0;  /* once become 1, remain 1. */
   lastmod = 0;   /* last modified line */
   crnt = NULL;

   // reads a line from fp and prints it
   while (!readln(fp, &line)) {
      newline_at_eol = lastch(line) == '\n';

      // with -l enabled, highlights whitespaces at eol
      if (of->eol) {
         int rv = handle_eoldrn(line, numline, of->lne);
         lret |= rv;
         if (rv) lastmod = numline;
      }
      else
      // otherwise, just prints the line untouched
      if (!of->lne)
         fmtwrt("%4d|%s", numline, line);

      prev = crnt;
      crnt = line;
      free(prev);
      numline++;
   }

   // crnt != NULL test is necessary because the file can be empty.
   if (crnt) {
      // with -f enabled, checks if there is a missing \n at eof
      // newline_at_eol must hold a value since crnt != NULL.
      if (of->eof) {
         fret = handle_eofdrn(numline, lastmod, of->lne, newline_at_eol);
         if (!lret && !fret)
            sputs("(Nothing to fix)");
      }
      else
      if (of->ful) {
         if (newline_at_eol)
            fmtwrt("%4d|\n", numline);
         else
            sputs("\n(No newline before EOF)");
      }
      else {   /* neither eof nor ful */
         if (!lret) sputs("(Nothing to fix)");
         else
         if (!newline_at_eol && lastmod == numline - 1)
            sputs("");
      }
   }
   else
   // empty file
   if (numline == 1 && !crnt)
      sputs("(EOF)");

   // do usual cleanup
   free(crnt);
}

// `cntspn` (count span) counts how many unnecessary spaces are at eol.
static int cntspn(char *line);
// `cntdgt` (count digit) counts how many digits are in `num`.
static unsigned int cntdgt(unsigned int num);

static int handle_eoldrn(char *line, int numline, bool dryrun_line) {
   unsigned int len, spn, cnt, ret;

   // calculates the amount of whitespace at eol
   len = strlen(line);
   spn = cntspn(line);
   cnt = cntdgt(spn);
   ret = 0;

   // whitespace being present at eol
   if (spn) {
      int nloc, nspace;
      bool cond;

      nloc = len - spn;
      cond = lastch(line) == '\n';
      if (cond) nloc--;
      line[nloc] = '\0';

      fmtwrt("%s%4d|%s%s%s",
         Cyellow, numline, Creset, line, CBred);

      nspace = spn - cnt;

      if (spn == 1)
         fmtwrt("1%s", Creset);
      else
         fmtwrt("%*c%d%s", nspace, ' ', spn, Creset);

      if (cond) sputs("");

      ret = 1;
   }
   else
   // no whitespace at eol, but needs to print the line unchanged
   if (!dryrun_line)
      fmtwrt("%4d|%s", numline, line);

   return ret;
}

static int cntspn(char *line) {
   int i, len, span;

   len = strlen(line);
   span = 0;

   i = len - 1;
   if (lastch(line) == '\n')
      i--;

   for (/* blank */; i >= 0; i--)
      if (match(line[i], " \f\t\v"))
         span++;
      else
         break;

   return span;
}

static unsigned int cntdgt(unsigned int num) {
   char buf[11]; /* 4,294,967,295 */
   int cnt;

   cnt = sprintf(buf, "%d", num);
   if (cnt < 1) vfatal("proofread: %s: buffer is not written.", __func__);

   return cnt;
}

static bool handle_eofdrn(int numline, int lastmod, bool dryrun_line, bool newline_at_eol) {
   bool ret = true;

   if (!newline_at_eol) {
      if (!dryrun_line || lastmod == numline - 1)
         sputs("");
      fmtwrt("%s%4d|%s%s\\n%s\n", Cyellow, numline, Creset, CBgreen, Creset);
   }
   else
   // already having \n at eof, just needs to print the line number
   if (!dryrun_line)
      fmtwrt("%4d|\n", numline);
   else
      ret = false;

   return ret;
}

// `apply_actual` performs a real task on a file based on `of`.
static void apply_actual(FILE *src, FILE *dest, optflg_t *of);
// `cleanup` removes and/or renames the original and result file.
static void cleanup(optflg_t *of, const char *srcname, const char *destname, const char *keeppath);

static void run_actual(optflg_t *of, dynarr_t *filenames, const char *keeppath) {
   // handles + option
   if (of->owf) {
      int idx, len;

      len = dynarr_len(filenames);
      for (idx = 0; idx < len; idx++) {
         char *srcname, *destname;
         FILE *src, *dest;

         // gets a stream
         srcname = *((char **) dynarr_get(filenames, idx));
         destname = concat(2, srcname, ".prfrd.tmp");
         src  = sfopen(srcname, "r");
         dest = sfopen(destname, "w");

         // main logic
         apply_actual(src, dest, of);

         // cleanup
         sfclose(src);
         sfclose(dest);
         cleanup(of, srcname, destname, keeppath);
         free(destname);
      }
   }
   // handles the default option
   else apply_actual(stdin, stdout, of);
}

// `handle_eolact` gives the eol-problem an actual fix.
static void handle_eolact(char *line, FILE *dest);
// `handle_eofact` gives the eof-problem an actual fix.
static void handle_eofact(FILE *dest, bool newline_at_eol);

static void apply_actual(FILE *src, FILE *dest, optflg_t *of) {
   char *line, *crnt, *prev;
   bool newline_at_eol;

   // reads a line from stdin and prints it
   crnt = NULL;

   while (!readln(src, &line)) {
      newline_at_eol = lastch(line) == '\n';

      // with -l enabled, removes whitespaces at eol
      if (of->eol)
         handle_eolact(line, dest);
      else
      // otherwise, just prints the line untouched
         sfputs(dest, line);
      prev = crnt;
      crnt = line;
      free(prev);
   }

   // with -f enabled, puts \n at eof, if not existed
   // crnt != NULL test is necessary because the file can be empty.
   if (of->eof && crnt)
      // newline_at_eol must hold a value since crnt != NULL.
      handle_eofact(dest, newline_at_eol);

   // do usual cleanup
   free(crnt);
}

static void handle_eolact(char *line, FILE *dest) {
   int len, span;
   bool cond;

   // calculates the amount of whitespace at eol
   len = strlen(line);
   span = cntspn(line);

   // whitespace being present at eol
   if (span) {
      cond = lastch(line) == '\n';
      if (cond) span++;
      line[len - span] = '\0';
      sfputs(dest, line);
      if (cond) sfputc(dest, '\n');
   }
   else
   // no whitespace at eol, but needs to print the line unchanged
      sfputs(dest, line);
}

static void handle_eofact(FILE *dest, bool newline_at_eol) {
   if (!newline_at_eol)
      sfputc(dest, '\n');
}

static void cleanup(optflg_t *of, const char *srcname, const char *destname, const char *keeppath) {
   #if !defined(_WIN32) && (defined(__unix__) || defined(__unix))
      char *dirsep = "/";
   #else
      char *dirsep = "\\";
   #endif

   #ifndef PATH
   #define PATH "bin"
   #endif

   char *rnmmsg = "%s is renamed to %s.\n";

   if (!of->kep) {
      /*
      srcfile  | input.txt -> (deleted)
      destfile | input.txt.prfrd.tmp -> input.txt
      */
      sremove(srcname);
      if (!of->mut) fmtwrt("%s is deleted.\n", srcname);
      srename(destname, srcname);
      if (!of->mut) fmtwrt(rnmmsg, destname, srcname);
      return;
   }
   else {
      /*
      srcfile  | d/input.txt -> keeppath/input.txt
      destfile | d/input.txt.prfrd.tmp -> d/input.txt
      */
      char *onlyname, *newname;

      /*
      No need to test onlyname == NULL since if so,
      the file wouldn't have opened in the first place.
      */
      onlyname = extfnm(srcname);
      if (!strcmp(keeppath, "bak"))
         newname = concat(5, PATH, dirsep, keeppath, dirsep, onlyname);
      else
         newname = concat(3, keeppath, dirsep, onlyname);

      if (rename(srcname, newname)) {
         sremove(destname);
         vfatal("proofread: unable to move %s to %s; "
            "maybe the given directory \"%s\" does not exist?",
            srcname, newname, keeppath);
      }
      if (!of->mut) fmtwrt(rnmmsg, srcname, newname);
      srename(destname, srcname);
      if (!of->mut) fmtwrt(rnmmsg, destname, srcname);
      free(newname);
      free(onlyname);
   }
}
